base/operators.jl

```diff
@@ -873,10 +873,28 @@ julia> ∘(fs...)(3)
 ```
 """
 function ∘ end
+
+struct ComposedFunction{F,G} <: Function
+    f::F
+    g::G
+    ComposedFunction{F, G}(f, g) where {F, G} = new{F, G}(f, g)
+    ComposedFunction(f, g) = new{Core.Typeof(f),Core.Typeof(g)}(f, g)
+end
+
+(c::ComposedFunction)(x...) = c.f(c.g(x...))
+
 ∘(f) = f
-∘(f, g) = (x...)->f(g(x...))
+∘(f, g) = ComposedFunction(f, g)
 ∘(f, g, h...) = ∘(f ∘ g, h...)
 
+function show(io::IO, c::ComposedFunction)
+    show(io, c.f)
+    print(io, " ∘ ")
+    show(io, c.g)
+end
+
+show(io::IO, ::MIME"text/plain", c::ComposedFunction) = show(io, c)
```

Comment:
There is already a fallback for this.

Reply:
Without this, text/plain output is rather ugly:

```
julia> uppercase∘first
(::Base.ComposedFunction{typeof(uppercase),typeof(first)}) (generic function with 1 method)
```

I think this is due to `show(io::IO, ::MIME"text/plain", f::Function)`.

But defining `show` here was incorrect because MIME is not defined here. I moved it to show.jl where `show(io::IO, ::MIME"text/plain", f::Function)` is defined.

Reply:
Oh, I see, I was missing that it's a subtype of Function.